feat(discover): improve invalid filter message

alxu-sentry opened the pull request:
There are a number of users at least on SaaS who are having difficulty
formatting their date-related filters. We should improve our docs and frontend
handling here, but I think this is a reasonable and very easy solution to
reduce the issues. I don't think we have room for a full exposition here, but a
brief description of the expected format seems reasonable.

Regarding the specific messages, not everybody knows what "ISO 8601" means, but
most developers should, and you can also just look it up.

alxu-sentry committed:
feat(discover): improve invalid filter message

There are a number of users at least on SaaS who are having difficulty
formatting their date-related filters. We should improve our docs and frontend
handling here, but I think this is a reasonable and very easy solution to
reduce the issues. I don't think we have room for a full exposition here, but a
brief description of the expected format seems reasonable.

Regarding the specific messages, not everybody knows what "ISO 8601" means, but
most developers should, and you can also just look it up.

wmak commented on src/sentry/api/event_search.py:
```diff
@@ -635,11 +635,17 @@ def _handle_basic_filter(self, search_key, operator, search_value):
         if search_key.name in self.date_keys:
-            raise InvalidSearchQuery(f"{search_key.name}: Invalid format for date field")
+            raise InvalidSearchQuery(
+                f"{search_key.name}: Invalid format for date field. Expected ISO 8601 or +/-duration."
```
Instead of making a user look up ISO8601, why not include an example? It'd also let the user copy and paste the example to fill in the field as well. We also already do this for an error message:
https://github.com/getsentry/sentry/blob/443efcca22e69723935923fcd0250ad30b546bd7/src/sentry/api/event_search.py#L1590

alxu-sentry replied:
Hm. I thought about that but thought it made the error message too long. Additionally, not all event_search usages allow the user to copy and paste. For example, the performance landing page only shows a toast which disappears when clicked on.

I think we can shorten "Invalid format for date field" to "Invalid date" though, which gives us more room for an example.

alxu-sentry committed:
make message more informative but still concise

Toast is totally broken on smartphone but the pages don't really work anyways.
The length works fine on Chrome's iPad simulation.

dashed commented on src/sentry/api/event_search.py:
```diff
@@ -635,11 +635,17 @@ def _handle_basic_filter(self, search_key, operator, search_value):
         if search_key.name in self.date_keys:
-            raise InvalidSearchQuery(f"{search_key.name}: Invalid format for date field")
+            raise InvalidSearchQuery(
+                f"{search_key.name}: Invalid date: {search_value.raw_value}. Expected +/-duration (e.g. +1h) or ISO 8601 (e.g. {datetime.now().isoformat()[:-4]})."
```
Is this usually RFC 3339?

alxu-sentry replied:
ok so after some research, our date format is a subset of RFC 3339 and ISO 8601.

dashed commented on src/sentry/api/event_search.py:
```diff
@@ -635,11 +635,17 @@ def _handle_basic_filter(self, search_key, operator, search_value):
         if self.is_numeric_key(search_key.name):
-            raise InvalidSearchQuery(f"{search_key.name}: Invalid format for numeric field")
+            raise InvalidSearchQuery(
+                f"{search_key.name}: Invalid number: {search_value.raw_value}. Expected number then optional k, m, or b, e.g. 500k."
```
```suggestion
                f"{search_key.name}: Invalid number: {search_value.raw_value}. Expected number followed by an optional prefix k, m, or b, e.g. 500k."
```

Zylphrex replied:
I think you mean `suffix` here.

dashed replied:
Oops. Yea suffix

alxu-sentry replied:
Adjusted in latest version. I don't want to make it too long, otherwise it overflows the toast.

alxu-sentry committed:
adjust message

alxu-sentry committed:
fixup tests

alxu-sentry committed:
update one more message and tests
